src/partials/lib/ForWorkersWidget/ForWorkersWidget.js

```diff
@@ -0,0 +1,125 @@
+    getWorkerTodoJobs: PropTypes.func.isRequired,
+    workerTodoJobsLoading: PropTypes.bool,
+    workerTodoJobsFailure: PropTypes.instanceOf(Error),
+    workerTodoJobs: PropTypes.arrayOf(PropTypes.object),
+    todoJobs: PropTypes.arrayOf(PropTypes.instanceOf(JobModel)),
+    todoDate:  PropTypes.string,
+  }
+
+  componentDidMount () {
+    this.props.getWorkerTodoJobs()
+  }
+
+  render () {
+    const { todoJobs, todoDate, signer } = this.props
+    return signer && (
+      <div className={css.main}>
+        <div className={css.row}>
+          <Widget
+            title='ui.dashboard.worker.startYourJobSearch'
+            subtitle='ui.dashboard.worker.worker'
+            actions={[
+              {
+                href: '/general-job-board',
+                label: 'Visit general job board',
+                isLink: true,
+              },
+              {
+                href: '/job-boards',
+                label: 'Browse job boards',
+                isLink: true,
+              },
+            ]}
+          >
+            You may visit our General Job Board and start your search or
+            browse Job Boards created by other network users.
+          </Widget>
+          <Widget
+            href='/opportunities'
+            title='ui.dashboard.worker.opportunities'
+            subtitle='ui.dashboard.worker.worker'
+            actions={[
+              {
+                href: '/opportunities',
+                label: 'Opportunities',
+                counter: { value: 135 },
+                isLink: true,
+              },
+              {
+                label: 'Offers',
+                counter: { value: 3 },
+                isLink: true,
+              },
+              {
+                label: 'Applications',
+                isLink: true,
+              },
+            ]}
+          />
+        </div>
+        { todoJobs.length > 0 ? (
+          <div className={css.row}>
+            <Widget
+              title='ui.dashboard.worker.toDo'
+              titlePlaceholder={todoDate}
+              subtitle='ui.dashboard.worker.worker'
+              actions={
+                todoJobs.map(job => ({
+                  label: job.ipfs.name,
+                  date: moment(job.ipfs.period.since).format('h:mm A'),
+                  isLink: true,
+                }))
+              }
+            />
+          </div>
+        ) : null }
+      </div>
+    )
+  }
+}
+
+const mapStateToProps = (state) => {
+  const workerTodoJobs = workerTodoJobsSelector(state)
+  let todoJobs = []
+  let todoDate = null
+  if (workerTodoJobs) {
+    workerTodoJobs.forEach(job => {
+      const date = job.ipfs.period.since
+      if (moment(date).isSame(todoDate, 'day')) {
+        todoJobs = todoJobs.concat(job)
+      } else if (moment().diff(date, 'days') < 0 && (moment(date).isBefore(todoDate, 'day') || todoDate === null)) {
+        todoDate = date
+        todoJobs = [job]
+      }
+    })
+  }
```

Comment:
maybe we can firstly find nearest date and then filter jobs by this date?
like this
```
const nearestDate = workerTodoJobs.sort((job) => job.ipfs.period.since)[0]
const nearestDateJobs = workerTodoJobs.filter((job) => job.ipfs.period.since === nearestDate)
```

Reply:
workerTodoJobs will be processed twice this way
Also we need not just earliest date but earliest date in future:
```
const nearestDate = workerTodoJobs.filter(job => new Date(job.ipfs.period.since) >= new Date()).sort((job) => job.ipfs.period.since)[0]
```
Also date should be not strict equal but same day:

```
const nearestDateJobs = workerTodoJobs.filter((job) => moment(job.ipfs.period.since).isSame(nearestDate, 'day'))
```
And .sort() will not work this way it needs to be rewritten to compare dates with *compareFunction*:
```
.sort((a, b) => (a.ipfs.period.since < b.ipfs.period.since ? -1 : 1))
```
So final solution is neither more readable nor compact in my opinion:
```
const nearestDate = workerTodoJobs
  .filter(job => new Date(job.ipfs.period.since) >= new Date())
  .sort((a, b) => (a.ipfs.period.since < b.ipfs.period.since ? -1 : 1))[0].ipfs.period.since
const nearestDateJobs= workerTodoJobs.filter((job) => moment(job.ipfs.period.since).isSame(todoDate, 'day'))
```

What do you think?

Reply:
ok 
in this case all is fine
but functional programming is preferable with chaining filter/map/reduce/ patterns

Reply:
Sure, I'm trying to escape such things when possible